File: ios/Classes/BlueDeviceInfoModel.m
#import "BlueDeviceInfoModel.h"

@implementation BlueDeviceInfoModel

- (instancetype)initWithDeviceName:(NSString *)deviceName deviceHardwareAddress:(NSString *)deviceHardwareAddress connectionState:(int)connectionState {
    self = [super init];
    if (self) {
        _deviceName = deviceName;
        _deviceHardwareAddress = deviceHardwareAddress;
        _connectionState = connectionState;
    }
    return self;
}

- (NSString *)toJson {
    NSError *error = nil;
    NSDictionary *dictionary = @{
        @"deviceName": self.deviceName,
        @"deviceHardwareAddress": self.deviceHardwareAddress,
        @"connectionState": @(self.connectionState)
    };
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dictionary options:NSJSONWritingPrettyPrinted error:&error];

    if (error) {
        return @"";
    } else {
        return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    }
}

- (BlueDeviceInfoModel *)fromJson:(NSString *)json {
    NSError *error = nil;
    NSData *jsonData = [json dataUsingEncoding:NSUTF8StringEncoding];
    NSDictionary *dictionary = [NSJSONSerialization JSONObjectWithData:jsonData options:kNilOptions error:&error];

    if (error) {
        return nil;
    } else {
        return [[BlueDeviceInfoModel alloc] initWithDeviceName:dictionary[@"deviceName"]
                                     deviceHardwareAddress:dictionary[@"deviceHardwareAddress"]
                                        connectionState:[dictionary[@"connectionState"] intValue]];
    }
}

@end

File: lib/model/blue_device_info_model.dart
class BlueDeviceInfoModel {
  BlueDeviceInfoModel({
    String? deviceName,
    String? deviceHardwareAddress,
    num? connectionState,
  }) {
    _deviceName = deviceName;
    _deviceHardwareAddress = deviceHardwareAddress;
    _connectionState = connectionState;
  }

  BlueDeviceInfoModel.fromJson(dynamic json) {
    _deviceName = json['deviceName'];
    _deviceHardwareAddress = json['deviceHardwareAddress'];
    _connectionState = json['connectionState'];
  }

  String? _deviceName;
  String? _deviceHardwareAddress;
  num? _connectionState;
  bool isLoading = false;

  BlueDeviceInfoModel copyWith({
    String? deviceName,
    String? deviceHardwareAddress,
    num? connectionState,
  }) =>
      BlueDeviceInfoModel(
        deviceName: deviceName ?? _deviceName,
        deviceHardwareAddress: deviceHardwareAddress ?? _deviceHardwareAddress,
        connectionState: connectionState ?? _connectionState,
      );

  String? get deviceName => _deviceName;

  String? get deviceHardwareAddress => _deviceHardwareAddress;

  num? get connectionState => _connectionState;

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{};
    map['deviceName'] = _deviceName;
    map['deviceHardwareAddress'] = _deviceHardwareAddress;
    map['connectionState'] = _connectionState;
    return map;
  }
}
